### PR TITLE
Firestore: Fix data race in credentials provider

### PR DESCRIPTION
**Pull request by dconeybe**
Fixes: #10393

### CODE REVIEW

File: Firestore/core/test/unit/credentials/firebase_auth_credentials_provider_test.mm [hunk 1]
```diff
@@ -153,6 +153,29 @@ - (void)getTokenForcingRefresh:(BOOL)forceRefresh
...
 
+// Catch race conditions in FirebaseAuthCredentialsProvider::GetToken() when
+// run under thread sanitizer.
+// See https://github.com/firebase/firebase-ios-sdk/issues/10393
+TEST(FirebaseAuthCredentialsProviderTest, GetTokenCalledByAnotherThread) {
+  FIRApp* app = testutil::AppForUnitTesting();
+  FSTAuthFake* auth = [[FSTAuthFake alloc] initWithToken:@"token for fake uid"
+                                                     uid:@"fake uid"];
+  FirebaseAuthCredentialsProvider credentials_provider(app, auth);
+  credentials_provider.InvalidateToken();
```
Comment (cherylEnkidu) [inline Firestore/core/test/unit/credentials/firebase_auth_credentials_provider_test.mm:164]: Could you please tell me why we need to call InvalidateToken()?

File: Firestore/core/test/unit/credentials/firebase_auth_credentials_provider_test.mm [hunk 2]
```diff
@@ -153,6 +153,29 @@ - (void)getTokenForcingRefresh:(BOOL)forceRefresh
...
+    credentials_provider.GetToken(
+        [](util::StatusOr<AuthToken> result) { EXPECT_TRUE(result.ok()); });
+  });
+  std::thread thread2([&credentials_provider] {
+    credentials_provider.GetToken(
+        [](util::StatusOr<AuthToken> result) { EXPECT_TRUE(result.ok()); });
+  });
+
+  thread1.join();
+  thread2.join();
```
Comment (cherylEnkidu) [inline Firestore/core/test/unit/credentials/firebase_auth_credentials_provider_test.mm:176]: What would happen to this test if we used previous nonatomic implementation?

**Review by cherylEnkidu — COMMENTED** [review]
Thanks for your fix! Just a few questions

File: Firestore/core/test/unit/credentials/firebase_auth_credentials_provider_test.mm (hunk 1 above)
Comment (dconeybe) [reply Firestore/core/test/unit/credentials/firebase_auth_credentials_provider_test.mm:164]: I honestly don't know. I just copied a test from above and tweaked it. If it blocks your code review, I can investigate further though.

Comment (cherylEnkidu) [reply Firestore/core/test/unit/credentials/firebase_auth_credentials_provider_test.mm:164]: Thank you!

File: Firestore/core/test/unit/credentials/firebase_auth_credentials_provider_test.mm (hunk 2 above)
Comment (dconeybe) [reply Firestore/core/test/unit/credentials/firebase_auth_credentials_provider_test.mm:176]: Changing `atomic` back to `nonatomic` causes the Thread Sanitizer tests to fail with a race condition. I don't recall exactly what the race condition was though. So, I've pushed up a commit to this PR, https://github.com/firebase/firebase-ios-sdk/pull/10403/commits/e9cc0f9a6d7c32ab14ce6a9927b1a3e06ca4cc18, that changes `atomic` back to `nonatomic`. We'll see how/if this test fails then I'll reply back to this comment.

File: Firestore/core/test/unit/credentials/firebase_auth_credentials_provider_test.mm (hunk 1 above)
Comment (dconeybe) [reply Firestore/core/test/unit/credentials/firebase_auth_credentials_provider_test.mm:164]: Actually... you're right. I had copied the code above, which is specifically testing InvalidateToken(). It probably isn't relevant for this test. I'll remove it.

**Review by cherylEnkidu — APPROVED** [review]
LGTM!

**Comment by tsahi-deri** [comment]
@dconeybe Hey, is this fix exist in release version 10.1.0?

**Comment by paulb777** [comment]
It will go into 10.2.0. I updated the milestone on the issue.

**Pull request merged**